api/errdefs/is.go

```diff
@@ -0,0 +1,153 @@
+		ConflictError,
+		UnauthorizedError,
+		UnavailableError,
+		ForbiddenError,
+		SystemError,
+		NotModifiedError,
+		NotImplementedError,
+		UnknownError:
+		return true
+	}
+	return false
+}
+
+// IsNotFound returns if the passed in error is a NotFoundError
+func IsNotFound(err error) bool {
+	if !isKnown(err) {
+		if causal, ok := err.(causer); ok {
+			return IsNotFound(causal.Cause())
+		}
+		return false
+	}
+
+	_, ok := err.(NotFoundError)
+	return ok
+}
+
+// IsInvalidParameter returns if the passed in error is an InvalidParameterError
+func IsInvalidParameter(err error) bool {
+	if !isKnown(err) {
+		if causal, ok := err.(causer); ok {
+			return IsInvalidParameter(causal.Cause())
+		}
+		return false
+	}
+
+	_, ok := err.(InvalidParameterError)
+	return ok
+}
+
+// IsConflict returns if the passed in error is a ConflictError
+func IsConflict(err error) bool {
+	if !isKnown(err) {
+		if causal, ok := err.(causer); ok {
+			return IsConflict(causal.Cause())
+		}
+		return false
+	}
+
+	_, ok := err.(ConflictError)
+	return ok
+}
+
+// IsUnauthorized returns if the the passed in error is an UnauthorizedError
+func IsUnauthorized(err error) bool {
+	if !isKnown(err) {
+		if causal, ok := err.(causer); ok {
+			return IsUnauthorized(causal.Cause())
+		}
+		return false
+	}
+
+	_, ok := err.(UnauthorizedError)
+	return ok
+}
+
+// IsUnavailable returns if the passed in error is an UnavailableError
+func IsUnavailable(err error) bool {
+	if !isKnown(err) {
+		if causal, ok := err.(causer); ok {
+			return IsUnavailable(causal.Cause())
+		}
+		return false
+	}
+
+	_, ok := err.(UnavailableError)
+	return ok
+}
+
+// IsForbidden returns if the passed in error is a ForbiddenError
+func IsForbidden(err error) bool {
+	if !isKnown(err) {
+		if causal, ok := err.(causer); ok {
+			return IsForbidden(causal.Cause())
+		}
+		return false
+	}
+
+	_, ok := err.(ForbiddenError)
+	return ok
+}
+
+// IsSystem returns if the passed in error is a SystemError
+func IsSystem(err error) bool {
+	if !isKnown(err) {
+		if causal, ok := err.(causer); ok {
+			return IsSystem(causal.Cause())
+		}
+		return false
+	}
+
+	_, ok := err.(SystemError)
+	return ok
+}
+
+// IsNotModified returns if the passed in error is a NotModified error
+func IsNotModified(err error) bool {
+	if !isKnown(err) {
+		if causal, ok := err.(causer); ok {
+			return IsNotModified(causal.Cause())
+		}
+		return false
+	}
+
+	_, ok := err.(NotModifiedError)
+	return ok
+}
+
+// IsNotImplemented returns if the passed in error is a NotImplementedError
+func IsNotImplemented(err error) bool {
+	if !isKnown(err) {
+		if causal, ok := err.(causer); ok {
+			return IsNotImplemented(causal.Cause())
+		}
+		return false
+	}
+
+	_, ok := err.(NotImplementedError)
+	return ok
+}
+
+// IsUnknown returns if the passed in error is an UnknownError
+func IsUnknown(err error) bool {
+	if !isKnown(err) {
+		if causal, ok := err.(causer); ok {
+			return IsNotImplemented(causal.Cause())
```

Comment:
Alternatively, it might be possible to simplify the code using reflection. The `Type` interface has an `Implements` method. There could be a generic function that recurses up the causal chain, and takes a `Type` to check the errors against. I'm not sure if this is worth it.